src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs

```diff
@@ -100,38 +102,69 @@ executionPayload switch
     public static int ExpectedEngineSpecVersion(this IReleaseSpec spec) =>
         spec switch
         {
-            { WithdrawalsEnabled: true, IsEip4844Enabled: true } => EngineApiVersions.Cancun,
+            { IsEip4844Enabled: true } => EngineApiVersions.Cancun,
             { WithdrawalsEnabled: true } => EngineApiVersions.Shanghai,
             _ => EngineApiVersions.Paris
         };
 
-    public static bool Validate(
-        this PayloadAttributes payloadAttributes,
-        IReleaseSpec spec,
-        int version,
+    public static PayloadAttributesValidationResult Validate(
+       this PayloadAttributes payloadAttributes,
+       ISpecProvider specProvider,
+       int apiVersion,
+       [NotNullWhen(false)] out string? error) =>
+        Validate(
+            apiVersion: apiVersion,
+            actualVersion: payloadAttributes.GetVersion(),
+            expectedVersion: specProvider.GetSpec(ForkActivation.TimestampOnly(payloadAttributes.Timestamp))
+                                         .ExpectedEngineSpecVersion(),
+            "PayloadAttributesV",
+            out error);
+
+    public static PayloadAttributesValidationResult Validate(
+        int apiVersion,
+        int actualVersion,
+        int expectedVersion,
+        string methodName,
         [NotNullWhen(false)] out string? error)
     {
-        int actualVersion = payloadAttributes.GetVersion();
-        int expectedVersion = spec.ExpectedEngineSpecVersion();
-
-        error = null;
-        if (actualVersion != expectedVersion)
+        if (apiVersion >= EngineApiVersions.Cancun)
         {
-            error = $"PayloadAttributesV{expectedVersion} expected";
+            if (actualVersion == apiVersion && expectedVersion != apiVersion)
+            {
+                error = $"{methodName}{expectedVersion} expected";
+                return PayloadAttributesValidationResult.UnsupportedFork;
+            }
         }
-        else if (actualVersion > version)
+        else if (apiVersion == EngineApiVersions.Shanghai)
```

Comment:
Can we write this in a generic way, without referencing explicit constants? That was the idea, now we have `EngineApiVersions.Shanghai` and `EngineApiVersions.Cancun` referenced everywhere and same problem on next hardfork, making code more and more complex.

Reply:
The code considers next forks, should work for any further one(for the future forks any API serves only one fork):
> if (apiVersion >= EngineApiVersions.Cancun)

Reply:
Maybe let's have `MaxFork` constant?

Reply:
I don't get it

Reply:
It works for new forks too, rules are just different for shanghai and cancun+further_forks.